scipy/stats/_binomtest.py

```diff
@@ -0,0 +1,258 @@
+            low, high = _binom_wilson_conf_int(self.k, self.n,
+                                               confidence_level,
+                                               self.alternative,
+                                               correction=method == 'wilsoncc')
+        return ConfidenceInterval(low=low, high=high)
+
+
+def _binom_exact_conf_int(k, n, confidence_level, alternative):
+    """
+    Compute the estimate and confidence interval for the binomial test.
+
+    Returns proportion, prop_low, prop_high
+    """
+    if alternative == 'two-sided':
+        alpha = (1 - confidence_level) / 2
+        if k == 0:
+            plow = 0.0
+        else:
+            plow = brentq(lambda p: binom.sf(k-1, n, p) - alpha, 0, 1)
+        if k == n:
+            phigh = 1.0
+        else:
+            phigh = brentq(lambda p: binom.cdf(k, n, p) - alpha, 0, 1)
+    elif alternative == 'less':
+        alpha = 1 - confidence_level
+        plow = 0.0
+        if k == n:
+            phigh = 1.0
+        else:
+            phigh = brentq(lambda p: binom.cdf(k, n, p) - alpha, 0, 1)
+    elif alternative == 'greater':
+        alpha = 1 - confidence_level
+        if k == 0:
+            plow = 0.0
+        else:
+            plow = brentq(lambda p: binom.sf(k-1, n, p) - alpha, 0, 1)
+        phigh = 1.0
+    return plow, phigh
+
+
+def _binom_wilson_conf_int(k, n, confidence_level, alternative, correction):
+    # This function assumes that the arguments have already been validated.
+    # In particular, `alternative` must be one of 'two-sided', 'less' or
+    # 'greater'.
+    p = k / n
+    if alternative == 'two-sided':
+        z = ndtri(0.5 + 0.5*confidence_level)
+    else:
+        z = ndtri(confidence_level)
+
+    t = 1 + z**2/n
+    r = (p + z**2/(2*n)) / t
+
+    if correction:
+        if alternative == 'less' or k == 0:
+            lo = 0.0
+        else:
+            dlo = ((z * sqrt(z**2 - 1/n + 4*n*p*(1 - p) + (4*p - 2)) + 1) /
+                   (2*n*t))
+            lo = r - dlo
+        if alternative == 'greater' or k == n:
+            hi = 1.0
+        else:
+            dhi = ((z * sqrt(z**2 - 1/n + 4*n*p*(1 - p) - (4*p - 2)) + 1) /
+                   (2*n*t))
+            hi = r + dhi
+    else:
+        d = z/t * sqrt(p*(1-p)/n + (z/(2*n))**2)
+        if alternative == 'less' or k == 0:
+            lo = 0.0
+        else:
+            lo = r - d
+        if alternative == 'greater' or k == n:
+            hi = 1.0
+        else:
+            hi = r + d
+
+    return lo, hi
+
+
+def binomtest(k, *, n=None, p=0.5, alternative='two-sided'):
+    """
+    Perform a test that the probability of success is p.
+
+    This is a test of the null hypothesis that the probability of success
+    in a Bernoulli experiment is `p`.
+
+    Parameters
+    ----------
+    k : int or array_like
+        The number of successes, or if k has length 2, it is the
+        number of successes and the number of failures.
```

Comment:
Agreed, there is no need to duplicate the old API, and I think accepting just the scalars `k` and `n` makes more sense.

The remaining parameters don't have to be keyword only.  One of the advantages of keyword-only parameters is that we are not forced to add future parameters at the end of the parameter list.

Reply:
But order doesn't matter much if they're all keyword-only, right?

Reply:
In my last update, I simplified the API so that `k` and `n` must be scalar integers.  (We can update the code to handle array inputs in the future if there is ever a demand for that.)

For the keyword only issue, let me make sure I understand the question.  By using the isolated `*` in the parameter list, all subsequent parameters must be specified by keyword when the function is called.  That means a user can't write `binomtest(3, 15, 0.25, 'greater')`; the user *must* write `binomtest(3, n=15, p=0.25, alternative='greater')` (but the order of the named parameters can be different).  If in the future we want to add the parameter `foo=None` (say) to the signature, we can put it anywhere after the `*` without breaking backwards compatibility.  That is, in the future we could update the function to

    def binomtest(k, *, n=None, p=0.5, foo=None, alternative='two-sided')

and we could be sure that there is no existing code of the form `binomtest(3, 15, 0.25, 'greater')` that would break with such a change.

Having said that, I'm not locked into using `*` as I've currently written it. Maybe the `*` should come after `n`, or even before `k`, ... or it could be dropped completely.  I think judicious use of `*` in the signature of functions with more than a few parameters makes it easier to expand the API in a backwards compatible way and forces the user to write more readable code, but I'm not sure how aggressively we want to adopt its use in our function signatures at the moment.  Comments from anyone would be welcome.

Reply:
I understand. `binom_test` is at least seven years old. I have a hard time seeing how any new parameters would take priority over `n`, which is required, and `p`, which is pretty fundamental. I could maybe see putting it after `p`, but not before `n`.

Reply:
I just pushed an update.  I removed the use of `*` from the signature, so none of the parameters *require* a keyword.  For this function, I'm OK with that, but if anyone feels strongly that we should put `*` in there somewhere, let me know.